test/snippet/search/bi_fm_index_cursor.cpp

```diff
@@ -13,106 +13,106 @@ int main()
-
-auto it = index.begin();         // create an cursor
-it.extend_right("AAC"_dna4);     // search the sequence "AAC"
-debug_stream << it.query() << '\n'; // outputs "AAC"
-it.extend_left("ATG"_dna4);      // extend the query to "ATGAAT"
-                                 // The rightmost character of "ATG" is extended to the left first.
-debug_stream << it.query() << '\n'; // outputs "ATGAAT"
+bi_fm_index<false> index{genome};          // build the bidirectional index
+
+auto cur = index.begin();                  // create a cursor
+cur.extend_right("AAC"_dna4);              // search the sequence "AAC"
+debug_stream << cur.query(genome) << '\n'; // outputs "AAC"
+cur.extend_left("ATG"_dna4);               // extend the query to "ATGAAT"
+                                           // The rightmost character of "ATG" is extended to the left first.
+debug_stream << cur.query(genome) << '\n'; // outputs "ATGAAT"
 //! [extend_left_seq]
 }
 
 {
 debug_stream << "Example cycle_back() and cycle_front()\n";
 //! [cycle]
 std::vector<dna4> genome{"GAATTAATGAAC"_dna4};
-bi_fm_index index{genome};           // build the index
-
-auto it = index.begin();             // create an cursor
-// it.cycle_back();                  // cycle_back / cycle_front on begin() is undefined behaviour!
-it.extend_right("AAC"_dna4);         // search the sequence "AAC"
-debug_stream << it.query() << '\n';     // outputs "AAC"
-debug_stream << it.last_char() << '\n'; // outputs 'C'
-
-// it.cycle_front();                 // undefined behaviour! only cycle_back() is allowed after extend_right()
-it.cycle_back();                     // search the sequence "AAT"
-debug_stream << it.query() << '\n';     // outputs "AAT"
-debug_stream << it.last_char() << '\n'; // outputs 'T'
-
-it.extend_left('G'_dna4);             // search the sequence "GAAT"
-debug_stream << it.query() << '\n';     // outputs "GAAC"
-debug_stream << it.last_char() << '\n'; // outputs 'G'
-
-// it.cycle_back();                  // undefined behaviour! only cycle_front() is allowed after extend_left()
-it.cycle_front();                    // search the sequence "TAAT"
-debug_stream << it.query() << '\n';     // outputs "TAAT"
-debug_stream << it.last_char() << '\n'; // outputs 'T'
-
-it.cycle_front();                    // search the sequence "TAAT"
-debug_stream << it.query() << '\n';     // outputs "TAAT"
-debug_stream << it.last_char() << '\n'; // outputs 'T'
+bi_fm_index<false> index{genome};                       // build the bidirectional index
+
+auto cur = index.begin();                               // create a cursor
+// cur.cycle_back();                                    // cycle_back / cycle_front on begin() is undefined behaviour!
+cur.extend_right("AAC"_dna4);                           // search the sequence "AAC"
+debug_stream << cur.query(genome) << '\n';              // outputs "AAC"
+debug_stream << cur.template last_char<dna4>() << '\n'; // outputs 'C'
```

Comment:
Why is the `last_char` function a template now? This doesn't look nice :see_no_evil: But the user does not really use the cursor does (s)he?

Reply:
The index/cursor don't know the text type, and work on ranks. So you need to provide the type if you want the last character. 
Usually, you would use the search interface.